File: Exam_Review/Algorithms/BFS.c
#include "Queue.h"
#include "Graph.h"
#include <stdio.h>
#include <stdlib.h>

bool BFS(Graph *graph, Node *start) {
    // Reset all nodes
    for (int i = 0; i < graph->node_count; i++) {
        graph->nodes[i].visited = 0;
        graph->nodes[i].distance_from_start = INFINITY;
    }

    // Stack initialization
    Queue *queue;
    init(&queue);
    enqueue(queue, start);
    start->visited = 1;
    start->distance_from_start = 0;

    while (!is_empty(queue)) {
        Node *cur = dequeue(queue);
        printf("Visited: %s\n", cur->name);

        for (int i = 0; i < cur->connection_count; i++) {
            Node *neighbor = cur->connections[i].node;
            if (!neighbor->visited) {
                neighbor->visited = 1;
                neighbor->distance_from_start = cur->distance_from_start + 1;
                enqueue(queue, neighbor);
            }
        }
    }

    free_queue(queue); // if implemented

    return true;
}

File: Exam_Review/Algorithms/BFS.py
from Queue import Queue
from Graph import Graph

def BFS(g, start):
    start_i = g.node_names(start)
    visited = [False] * g.num_nodes
    Q = Queue()
    Q.enqueue(start_i)

    while not Q.is_empty():
        node_index = Q.dequeue()
        
        if not visited[node_index]:
            visited[node_index] = True

            neighbor = g.nodes[node_index].head
            while neighbor:
                if not visited[neighbor.data]:
                    Q.enqueue(neighbor.data)
                neighbor = neighbor.next
